concordium-consensus/src/Concordium/MultiVersion.hs

```diff
@@ -999,48 +999,79 @@ sendCatchUpStatus genIndex = MVR $ \mvr@MultiVersionRunner{..} -> do
 -- During a protocol update, it might be wise to cache messages for
 -- the forthcoming protocol version. For the time being, we will
 -- simply rely on catch-up.
-withLatestExpectedVersion ::
+withLatestExpectedVersion' ::
     GenesisIndex ->
-    (EVersionedConfiguration gsconf finconf -> MVR gsconf finconf UpdateResult) ->
-    MVR gsconf finconf UpdateResult
-withLatestExpectedVersion gi a = do
+    (EVersionedConfiguration gsconf finconf -> MVR gsconf finconf (Skov.UpdateResult, Maybe a)) ->
+    MVR gsconf finconf (Skov.UpdateResult, Maybe a)
+withLatestExpectedVersion' gi a = do
     vvec <- liftIO . readIORef =<< asks mvVersions
     -- Length is an Int and GenesisIndex is a Word32.
     -- Assuming a 64-bit system, there is no risk of over/underflow.
     case compare (Vec.length vvec - 1) (fromIntegral gi) of
         EQ -> a (Vec.last vvec)
-        LT -> return ResultInvalidGenesisIndex
-        GT -> return ResultConsensusShutDown
+        LT -> return (Skov.ResultInvalidGenesisIndex, Nothing)
+        GT -> return (Skov.ResultConsensusShutDown, Nothing)
+
+withLatestExpectedVersion ::
+    GenesisIndex ->
+    (EVersionedConfiguration gsconf finconf -> MVR gsconf finconf Skov.UpdateResult) ->
+    MVR gsconf finconf Skov.UpdateResult
+withLatestExpectedVersion gi a = fst <$> withLatestExpectedVersion' gi (fmap (,Nothing) <$> a)
+
+-- |A continuation for executing a block that has been received
+-- and verified.
+newtype ExecuteBlock = ExecuteBlock {runBlock :: IO Skov.UpdateResult}
```

Comment:
What is the reasoning behind us not maintaining the lock through FFI call?

For example, we have looked up the parent and resolved some things already in the receiveBlock call. Then we repeat some of this in the executeBlock.

Now if we do not retain the lock through that then what we do now is necessary. But an alternative would have been to really just split the old receiveBlock, including retaining the state lock throughout.

I don't think you need to change it since it is done now, but I would like to understand the reasoning behind this choice. What tradeoffs were considered?

Reply:
Reasoning etc. are discussed here: https://github.com/Concordium/concordium-node/pull/611#discussion_r1031517447
So resolving this.